### PR TITLE
Wdp 116

### PR DESCRIPTION
**Pull request by gonczor**
Added tests, updated .gitignore for coverage files, updated requirements.txt

### CODE REVIEW

**Comment by gonczor** [comment]
image

https://kindergarten.stxnext.pl/secure/RapidBoard.jspa?rapidView=17&view=detail&selectedIssue=WDP-116

**Comment by gonczor** [comment]
image

File: runtime/data/test_data_with_header_and_footer.csv [hunk 1]
```diff
@@ -0,0 +1,11 @@
+header
```
Comment (dpolcyn) [inline runtime/data/test_data_with_header_and_footer.csv:1]: please explain me this 'header' thing, also why did you created new file with test data?

File: src/presence_analyzer/tests.py [hunk 2]
```diff
@@ -53,6 +62,67 @@ def test_api_users(self):
         self.assertEqual(len(data), 2)
         self.assertDictEqual(data[0], {u'user_id': 10, u'name': u'User 10'})
 
+    def test_presence_weekday_wrong_id(self):
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: wrong id? what is wrong with 1?

File: src/presence_analyzer/tests.py [hunk 3]
```diff
@@ -53,6 +62,67 @@ def test_api_users(self):
...
+        """
+        Test data contains ids: 10 and 11. Giving number -1 should result in 404 exit code
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: -1? why not just id of user that doesn't exist?

File: src/presence_analyzer/tests.py [hunk 4]
```diff
@@ -53,6 +62,67 @@ def test_api_users(self):
...
+        Test status code is 404
+        """
+        response = self.client.get('/api/v1/presence_weekday/1')
+        self.assertEqual(response.status_code, 404)
+
+    def test_presence_weekday_view(self):
+        """
+        Test json response for user 10
+        From test_data.csv:
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: In which standard doc did you wrote this? Link please

File: src/presence_analyzer/tests.py [hunk 5]
```diff
@@ -53,6 +62,67 @@ def test_api_users(self):
...
+        2013-09-10: Tuesday 30047 seconds
+        2013-09-11: Wednesday 24465 seconds
+        2013-09-12: Thursday 23705 seconds
+        """
+        response = self.client.get('/api/v1/presence_weekday/10')
+        self.assertEqual(response.status_code, 200)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Empty line after this line

File: src/presence_analyzer/tests.py [hunk 6]
```diff
@@ -53,6 +62,67 @@ def test_api_users(self):
...
+        data = json.loads(response.data)
+        # Change to dict for easier checks
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: check json loads format

File: src/presence_analyzer/tests.py [hunk 7]
```diff
@@ -53,6 +62,67 @@ def test_api_users(self):
...
+        data = dict(data)
+        self.assertEqual(len(data), 8)
+        self.assertEqual(data[u'Weekday'], u'Presence (s)')
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: ```from __future__ import unicode_literals```

File: src/presence_analyzer/tests.py [hunk 8]
```diff
@@ -53,6 +62,67 @@ def test_api_users(self):
...
+        self.assertEqual(data[u'Mon'], 0)
+        self.assertEqual(data[u'Tue'], 30047)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: If you gona check for example year, then you gona check every day too? it should be just dict check or list check, you don't have to check all position one by one

File: src/presence_analyzer/tests.py [hunk 9]
```diff
@@ -53,6 +62,67 @@ def test_api_users(self):
...
+        2013-09-11: Wednesday 24465 seconds
+        2013-09-12: Thursday 23705 seconds
+        """
+        response = self.client.get('/api/v1/mean_time_weekday/10')
+        self.assertEqual(response.status_code, 200)
+        data = json.loads(response.data)
+        # Change to dict for easier checks
+        data = dict(data)
+        self.assertEqual(len(data), 7)
+        self.assertEqual(data[u'Mon'], 0)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: ditto

File: src/presence_analyzer/tests.py [hunk 10]
```diff
@@ -53,6 +62,67 @@ def test_api_users(self):
...
+        From test_data.csv:
+        2013-09-10: Tuesday 30047 seconds
+        2013-09-11: Wednesday 24465 seconds
+        2013-09-12: Thursday 23705 seconds
+        """
+        response = self.client.get('/api/v1/mean_time_weekday/10')
+        self.assertEqual(response.status_code, 200)
+        data = json.loads(response.data)
+        # Change to dict for easier checks
+        data = dict(data)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: ditto

File: src/presence_analyzer/tests.py [hunk 11]
```diff
@@ -53,6 +62,67 @@ def test_api_users(self):
...
+        self.assertEqual(len(data), 7)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: if you check whole dict then you don't have to check length

File: src/presence_analyzer/tests.py [hunk 12]
```diff
@@ -86,8 +156,89 @@ def test_get_data(self):
...
 
+    def test_seconds_since_midnight(self):
+        """
+        Test time elapsed with different time set. Special case: midnight.
+        """
+        midnight_time = datetime.time(0, 0, 0)
+        second_after_midnight = datetime.time(0, 0, 1)
+        minute_after_midnight = datetime.time(0, 1, 0)
+        hour_after_midnight = datetime.time(1, 0, 0)
+        self.assertEqual(utils.seconds_since_midnight(midnight_time), 0)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: don't assign variable if u want to use it once

File: src/presence_analyzer/tests.py [hunk 13]
```diff
@@ -86,8 +156,89 @@ def test_get_data(self):
...
+        self.assertEqual(utils.seconds_since_midnight(second_after_midnight), 1)
+        self.assertEqual(utils.seconds_since_midnight(minute_after_midnight), 60)
+        self.assertEqual(utils.seconds_since_midnight(hour_after_midnight), 3600)
+
+    def test_interval(self):
+        """
+        Test interval
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: more information, full stop on the end of docstring

File: src/presence_analyzer/tests.py [hunk 14]
```diff
@@ -86,8 +156,89 @@ def test_get_data(self):
...
+        Mean checks whether provided list is empty and returns 0 if so.
+        """
+        self.assertEqual(utils.mean([]), 0)
+
+    def test_mean_with_non_empty_list(self):
+        """
+        Count mean and assert it provides 3-point accuracy.
+        """
+        items = [1, 2]
+        self.assertAlmostEqual(utils.mean(items), 1.5, 3)
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: Do both in one test, dont assign var if ur using it only once.

File: src/presence_analyzer/tests.py [hunk 15]
```diff
@@ -86,8 +156,89 @@ def test_get_data(self):
...
+        }
+
+        weekday_result = utils.group_by_weekday(data)
+        # From Monday to Saturday no data has been entered, so there should be
+        # empty lists
+        for i in range(6):
+            self.assertEqual(len(weekday_result[i]), 0)
+        # Ensure there is one entry for Sunday
+        self.assertEqual(len(weekday_result[6]), 1)
+        # Ensure it contains 1 hour interval
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: Imo ur overusing comments.
Compare whole dict instead of all this?

File: src/presence_analyzer/tests.py [hunk 16]
```diff
@@ -86,8 +156,89 @@ def test_get_data(self):
...
+        with testfixtures.LogCapture() as log_capture:
+
+            main.app.config.update({'DATA_CSV': INVALID_FORMAT_TEST_DATA})
+            try:
+                utils.get_data()
+            except UnboundLocalError:
+                pass
+            main.app.config.update({'DATA_CSV': TEST_DATA_CSV})
+            self.assertTrue(any(filter(lambda x:
+                                       'Problem with line' in x.msg, log_capture.records)))
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: I don't like this + wrong styles. If u rly wanted to test it u should use mock like:

image

File: src/presence_analyzer/tests.py [hunk 17]
```diff
@@ -86,8 +156,89 @@ def test_get_data(self):
...
+        Captures log and checks whether 'Problem with line' message is found.
+        """
+        with testfixtures.LogCapture() as log_capture:
+
+            main.app.config.update({'DATA_CSV': INVALID_FORMAT_TEST_DATA})
+            try:
+                utils.get_data()
+            except UnboundLocalError:
+                pass
+            main.app.config.update({'DATA_CSV': TEST_DATA_CSV})
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: This is not needed. Check what `setUp` func does

**Review by piotrgrundas — CHANGES_REQUESTED** [review]

File: runtime/data/test_data_with_header_and_footer.csv (hunk 1 above)
Comment (gonczor) [reply runtime/data/test_data_with_header_and_footer.csv:1]: If there is a different number of fields than 4, get_data() skips it. The comment said it was for checking for headers and footers, so I've added a file with data for testing this piece of code. I didn't want to mess with test_data.csv

File: src/presence_analyzer/tests.py (hunk 2 above)
Comment (gonczor) [reply src/presence_analyzer/tests.py]: It's not in test_data.csv, so the user doesn't exist. I guess I should make this more descriptive?

File: src/presence_analyzer/tests.py (hunk 3 above)
Comment (gonczor) [reply src/presence_analyzer/tests.py]: Done.

File: src/presence_analyzer/tests.py (hunk 5 above)
Comment (gonczor) [reply src/presence_analyzer/tests.py]: Done

File: src/presence_analyzer/tests.py (hunk 7 above)
Comment (gonczor) [reply src/presence_analyzer/tests.py]: Done

File: src/presence_analyzer/tests.py (hunk 12 above)
Comment (gonczor) [reply src/presence_analyzer/tests.py]: Done

File: src/presence_analyzer/tests.py (hunk 13 above)
Comment (gonczor) [reply src/presence_analyzer/tests.py]: Done

File: src/presence_analyzer/tests.py (hunk 14 above)
Comment (gonczor) [reply src/presence_analyzer/tests.py]: Done

File: src/presence_analyzer/tests.py (hunk 17 above)
Comment (gonczor) [reply src/presence_analyzer/tests.py]: Done

File: src/presence_analyzer/tests.py (hunk 16 above)
Comment (gonczor) [reply src/presence_analyzer/tests.py]: I couldn't get mock.patch to work as I wanted. I can check contents of log_capture instead.

File: src/presence_analyzer/tests.py (hunk 15 above)
Comment (gonczor) [reply src/presence_analyzer/tests.py]: Yep, cleared.

File: src/presence_analyzer/tests.py (hunk 4 above)
Comment (gonczor) [reply src/presence_analyzer/tests.py]: Corrected

File: src/presence_analyzer/tests.py (hunk 11 above)
Comment (gonczor) [reply src/presence_analyzer/tests.py]: Done

File: src/presence_analyzer/tests.py (hunk 8 above)
Comment (gonczor) [reply src/presence_analyzer/tests.py]: Done

File: src/presence_analyzer/utils.py [hunk 18]
```diff
@@ -65,6 +65,7 @@ def get_data():
                 end = datetime.strptime(row[3], '%H:%M:%S').time()
             except (ValueError, TypeError):
                 log.debug('Problem with line %d: ', i, exc_info=True)
+                # logging.debug('Sample msg')
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: rly?

File: src/presence_analyzer/tests.py [hunk 19]
```diff
@@ -86,8 +144,73 @@ def test_get_data(self):
...
+        """
+        data = {
+            datetime.date(2017, 4, 9): {
+                'start': datetime.time(12, 0, 0),
+                'end': datetime.time(13, 0, 0)
+            }
+        }
+        expected_data = [[], [], [], [], [], [], [3600]]
+
+        data = utils.group_by_weekday(data)
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py:211]: vars

    post/some work etc

    asserts

File: src/presence_analyzer/tests.py [hunk 20]
```diff
@@ -86,8 +144,73 @@ def test_get_data(self):
...
+        """
+        Mean checks whether provided list is empty and returns 0 if so.
+        """
+        self.assertEqual(utils.mean([]), 0)
+
+    def test_mean_with_non_empty_list(self):
+        """
+        Count mean and assert it provides 3-point accuracy.
+        """
+        self.assertAlmostEqual(utils.mean([1, 2]), 1.5, 3)
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: `test_mean_with_empty_list` and `test_mean_with_non_empty_list` in one test
Couple more cases.

File: runtime/data/test_data_with_header_and_footer.csv (hunk 1 above)
Comment (piotrgrundas) [inline runtime/data/test_data_with_header_and_footer.csv]: You won't have headers like that, in csv in 99% cases num of header cols will match data num cols.

File: src/presence_analyzer/tests.py [hunk 21]
```diff
@@ -2,18 +2,29 @@
...
 Presence analyzer unit tests.
 """
+from __future__ import unicode_literals
+
+import calendar
 import os.path
 import json
 import datetime
 import unittest
+from mock import patch
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: sort imports + group them
https://www.python.org/dev/peps/pep-0008/#imports

File: src/presence_analyzer/tests.py [hunk 22]
```diff
@@ -48,11 +59,58 @@ def test_api_users(self):
...
+        """
+        response = self.client.get('/api/v1/presence_weekday/10')
+        self.assertEqual(response.status_code, 200)
+        data = json.loads(response.data)
+
+        weekdays = utils.group_by_weekday(utils.get_data()[10])
+        expected_data = [
+                [calendar.day_abbr[weekday], sum(intervals)]
+                for weekday, intervals in enumerate(weekdays)
+            ]
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: wrong indent - untab once

File: src/presence_analyzer/tests.py [hunk 23]
```diff
@@ -48,11 +59,58 @@ def test_api_users(self):
...
+        """
+        response = self.client.get('/api/v1/mean_time_weekday/10')
+        self.assertEqual(response.status_code, 200)
+
+        data = json.loads(response.data)
+        weekdays = utils.group_by_weekday(utils.get_data()[10])
+        expected_data = [
+                [calendar.day_abbr[weekday], utils.mean(intervals)]
+                for weekday, intervals in enumerate(weekdays)
+            ]
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: ditto

File: src/presence_analyzer/tests.py [hunk 24]
```diff
@@ -86,8 +144,73 @@ def test_get_data(self):
...
+        self.assertEqual(utils.seconds_since_midnight(datetime.time(0, 0, 1)), 1)
+        self.assertEqual(utils.seconds_since_midnight(datetime.time(0, 1, 0)), 60)
+        self.assertEqual(utils.seconds_since_midnight(datetime.time(1, 0, 0)), 3600)
+
+    def test_interval(self):
+        """
+        Test interval is counted correctly for two different and two same points in time.
+        """
+        self.assertEqual(utils.interval(datetime.time(0, 0, 0), datetime.time(0, 0, 0)), 0)
+        self.assertEqual(utils.interval(datetime.time(0, 0, 0), datetime.time(1, 1, 1)), 3661)
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: Couple more cases.

File: src/presence_analyzer/tests.py [hunk 25]
```diff
@@ -48,11 +59,58 @@ def test_api_users(self):
...
+        self.assertEqual(response.status_code, 404)
+
+    def test_presence_weekday_view(self):
+        """
+        Test json response for user 10.
+        """
+        response = self.client.get('/api/v1/presence_weekday/10')
+        self.assertEqual(response.status_code, 200)
+        data = json.loads(response.data)
+
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: image

**Review by piotrgrundas — CHANGES_REQUESTED** [review]

File: src/presence_analyzer/tests.py [hunk 26]
```diff
@@ -86,8 +144,73 @@ def test_get_data(self):
...
+        """
+        Count mean and assert it provides 3-point accuracy.
+        """
+        self.assertAlmostEqual(utils.mean([1, 2]), 1.5, 3)
+
+    @patch('presence_analyzer.utils.log')
+    def test_invalid_data_handled(self, mocked_log):
+        """
+        On introduction of invalid data logger with message 'Problem with line 0' should be called.
+        :param mocked_log: mocked logger.
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: IMO adding :param to tests is pointless.

**Review by piotrgrundas — CHANGES_REQUESTED** [review]

File: src/presence_analyzer/utils.py (hunk 18 above)
Comment (gonczor) [reply src/presence_analyzer/utils.py]: Sorry, I forgot to delete it after testing logging mock.

File: src/presence_analyzer/tests.py [hunk 27]
```diff
@@ -48,11 +60,60 @@ def test_api_users(self):
...
+        Test json response for user 10.
+        """
+        response = self.client.get('/api/v1/mean_time_weekday/10')
+
+        data = json.loads(response.data)
+        weekdays = utils.group_by_weekday(utils.get_data()[10])
+        expected_data = [
+            [calendar.day_abbr[weekday], utils.mean(intervals)]
+            for weekday, intervals in enumerate(weekdays)
+            ]
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: expected_data = [
        [calendar.day_abbr[weekday], utils.mean(intervals)]
        for weekday, intervals in enumerate(weekdays)
    ]

**Review by piotrgrundas — CHANGES_REQUESTED** [review]

**Review by piotrgrundas — APPROVED** [review]

**Pull request merged**